tests/contrib/io/matplotlib/test_matplotlib_s3_writer.py

```diff
@@ -0,0 +1,238 @@
+    with mock_s3():
+        conn = s3fs.core.boto3.client("s3", **AWS_CREDENTIALS)
+        conn.create_bucket(Bucket=BUCKET_NAME)
+        yield conn
+
+
+@pytest.fixture
+def mocked_encrypted_s3_bucket():
+    bucket_policy = {
+        "Version": "2012-10-17",
+        "Id": "PutObjPolicy",
+        "Statement": [
+            {
+                "Sid": "DenyUnEncryptedObjectUploads",
+                "Effect": "Deny",
+                "Principal": "*",
+                "Action": "s3:PutObject",
+                "Resource": "arn:aws:s3:::{}/*".format(BUCKET_NAME),
+                "Condition": {"Null": {"s3:x-amz-server-side-encryption": "aws:kms"}},
+            }
+        ],
+    }
+    bucket_policy = json.dumps(bucket_policy)
+
+    with mock_s3():
+        conn = s3fs.core.boto3.client("s3", **AWS_CREDENTIALS)
+        conn.create_bucket(Bucket=BUCKET_NAME)
+        conn.put_bucket_policy(Bucket=BUCKET_NAME, Policy=bucket_policy)
+
+        yield conn
+
+
+@pytest.fixture
+def plot_writer(mocked_s3_bucket):  # pylint: disable=unused-argument
+    def _matplotlibwriters3():
+        return MatplotlibS3Writer(
+            bucket_name=BUCKET_NAME, filepath=KEY_PATH, credentials=AWS_CREDENTIALS
+        )
+
+    return _matplotlibwriters3()
+
+
+def test_save_data(tmp_path, mock_single_plot, plot_writer, mocked_s3_bucket):
+    """Test saving single matplotlib plot to S3."""
+
+    plot_writer.save(mock_single_plot)
+
+    expected_path = tmp_path / "downloaded_image.png"
+    actual_filepath = tmp_path / "locally_saved.png"
+
+    plt.savefig(str(actual_filepath))
+
+    mocked_s3_bucket.download_file(BUCKET_NAME, KEY_PATH, str(expected_path))
+
+    assert actual_filepath.read_bytes() == expected_path.read_bytes()
+
+
+def test_list_save(tmp_path, mock_list_plot, plot_writer, mocked_s3_bucket):
+    """Test saving list of plots to S3."""
+
+    plot_writer.save(mock_list_plot)
+
+    for index in range(5):
+
+        expected_path = tmp_path / "downloaded_image.png"
+        actual_filepath = tmp_path / "locally_saved.png"
+
+        mock_list_plot[index].savefig(str(actual_filepath))
+
+        _key_path = KEY_PATH + "/" + str(index) + ".png"
+
+        mocked_s3_bucket.download_file(BUCKET_NAME, _key_path, str(expected_path))
+
+        assert actual_filepath.read_bytes() == expected_path.read_bytes()
+
+
+def test_dict_save(tmp_path, mock_dict_plot, plot_writer, mocked_s3_bucket):
+    """Test saving dictionary of plots to S3."""
+
+    plot_writer.save(mock_dict_plot)
+
+    for colour in COLOUR_LIST:
+
+        expected_path = tmp_path / "downloaded_image.png"
+        actual_filepath = tmp_path / "locally_saved.png"
+
+        mock_dict_plot[colour].savefig(str(actual_filepath))
+
+        _key_path = KEY_PATH + "/" + colour
+
+        mocked_s3_bucket.download_file(BUCKET_NAME, _key_path, str(expected_path))
+
+        assert actual_filepath.read_bytes() == expected_path.read_bytes()
+
+
+def test_bad_credentials(mock_dict_plot):
```

Comment:
Shouldn't this fail at the construction because `bad_writer.save` does not use credentials?

Reply:
I guess the credentials are only used when you try to write to s3, not during class construction?